dtfs-central-api/src/v1/controllers/utilisation-report-service/post-add-payment.controller/helpers.test.ts

```diff
@@ -0,0 +1,206 @@
+import { TfmSessionUser } from '../../../../types/tfm/tfm-session-user';
+import { aTfmSessionUser } from '../../../../../test-helpers/test-data/tfm-session-user';
+import { getMockQueryRunner } from '../../../../../test-helpers/mock-query-runner';
+import { FeeRecordRepo } from '../../../../repositories/fee-record-repo';
+import { NewPaymentDetails } from '../../../../types/utilisation-reports';
+
+describe('post-add-payment.controller helpers', () => {
+  describe('addPaymentToUtilisationReport', () => {
+    const reportId = 1;
+    const feeRecordIds = [1, 2, 3];
+    const paymentCurrency: Currency = 'GBP';
+
+    const tfmUser: TfmSessionUser = {
+      ...aTfmSessionUser(),
+      _id: new ObjectId().toString(),
+    };
+    const tfmUserId = tfmUser._id;
+
+    const newPaymentDetails: NewPaymentDetails = {
+      currency: paymentCurrency,
+      amountReceived: 100,
+      dateReceived: new Date(),
+      paymentReference: 'A payment reference',
+    };
+
+    const utilisationReport = UtilisationReportEntityMockBuilder.forStatus('RECONCILIATION_IN_PROGRESS').withId(reportId).build();
+    const utilisationReportStateMachine = UtilisationReportStateMachine.forReport(utilisationReport);
+
+    const feeRecordsInPaymentCurrency = feeRecordIds.map((feeRecordId) =>
+      FeeRecordEntityMockBuilder.forReport(utilisationReport).withId(feeRecordId).withPaymentCurrency(paymentCurrency).build(),
+    );
+
+    const mockQueryRunner = getMockQueryRunner();
+    const feeRecordFindBySpy = jest.spyOn(FeeRecordRepo, 'findBy');
+    const createQueryRunnerSpy = jest.spyOn(SqlDbDataSource, 'createQueryRunner');
+
+    const utilisationReportStateMachineConstructorSpy = jest.spyOn(UtilisationReportStateMachine, 'forReportId');
+    const handleEventSpy = jest.spyOn(utilisationReportStateMachine, 'handleEvent');
+
+    beforeEach(() => {
+      feeRecordFindBySpy.mockResolvedValue(feeRecordsInPaymentCurrency);
+      createQueryRunnerSpy.mockReturnValue(mockQueryRunner);
+      utilisationReportStateMachineConstructorSpy.mockResolvedValue(utilisationReportStateMachine);
+      handleEventSpy.mockResolvedValue(utilisationReport);
+    });
+
+    afterEach(() => {
+      jest.resetAllMocks();
+    });
+
+    it("calls the 'UtilisationReportStateMachine.forReportId' method with the supplied report id", async () => {
+      // Act
+      await addPaymentToUtilisationReport(reportId, feeRecordIds, tfmUser, newPaymentDetails);
+
+      // Assert
+      expect(utilisationReportStateMachineConstructorSpy).toHaveBeenCalledWith(reportId);
+    });
+
+    it("calls the 'FeeRecordRepo.findBy' method with the supplied list of fee record ids", async () => {
+      // Act
+      await addPaymentToUtilisationReport(reportId, feeRecordIds, tfmUser, newPaymentDetails);
+
+      // Assert
+      expect(feeRecordFindBySpy).toHaveBeenCalledWith({ id: In(feeRecordIds) });
+    });
```

Comment:
Test names which essentially say 'calls the function named blah' are very easy to get out of date if we update the function name or if actually end up fetching the required data a different way. Test names and assertions should focus on what the function is actually intending to do. This could be 'fetches the fee records corresponding to the payment' but I think perhaps more interesting to assert the parameters in a test that is checking we make use of the returned fee records correctly, or make use of the returned state machine correctly. 

One option also is that there is actually a neat way to achieve checking we call with the correct a parameters without having to explicitly assert it is to use this sort of pattern:
```
const fn = jest.fn();
when(fn).calledWith(1).mockReturnValue('yay!');
```
(you do then have to be careful the same response won't be sent for other parameter values though when you set up your mocks)

Reply:
I suppose the use of `jest-when` implicitly performs this test, like you say. I'm tempted to try using that for future tests, but seeing as I have included quite extensive explicit checks here I'd prefer to keep them this time and just update the test names (I may try using `when` for the state machine changes you've suggested below though)